# Introducción al Análisis de Datos con Python


# 3. Indexación y selección de dataframes

En este notebook, exploraremos un conjunto de datos real utilizando la biblioteca de Python **Pandas**. Además, aprenderemos a seleccionar un subconjunto de datos de interés. A menudo, los conjuntos de datos son muy grandes y contienen mucha información que no es relevante para nuestra pregunta de investigación. Por ello, es un paso fundamental en el análisis de datos es seleccionar un subconjunto de datos que contenga solo la información que nos interesa. Esto no solo hace que el análisis sea más rápido, sino que también nos permite concentrarnos en los datos que son realmente importantes para nuestro objetivo.

Trabajaremos con las **[Estadísticas de Defunciones Registradas (EDR) de 2020](https://www.inegi.org.mx/programas/edr/)**, una base de datos generada por el Instituto Nacional de Estadística y Geografía (INEGI) de México.



## Familiarizarse con los datos

Antes de escribir cualquier línea de código, es crucial familiarizarnos con los archivos que descargamos. Este paso nos ayuda a entender el tipo de datos que tenemos y a anticipar los desafíos que podríamos enfrentar.

En la [sección anterior](https://drive.google.com/file/d/1jHpCfFSxAwFHpT5pqZqHvalOcO_X1IIp/view?usp=sharing) obtuvimos los datos del [EDR del 2020](https://www.inegi.org.mx/programas/edr/#datos_abiertos). Los datos del EDR son una carpeta que contiene varias subcarpetas con archivos `.CSV`. Esto es común, ya que los proyectos de datos a menudo incluyen la base de datos principal, catálogos (o tablas auxiliares) y metadatos.

Revisemos la lista de carpetas para entender su propósito:

- `conjunto_de_datos`: Este es el **conjunto de datos principal**. Contiene la información de cada defunción registrada en 2020. Aquí es donde encontraremos los datos que analizaremos, además de la bitacora de cambios y notas al respecto.
- `diccionario_de_datos` El **diccionario de datos** es el documento más importante en esta etapa, ya que describe el contenido de la base de datos, los campos y los catálogos.
- `metadatos`: Los **metadatos** son los datos sobre los datos, es decir, la descripción del conjunto de datos y como fue generada.
- `catalogos`: Estos son los **catálogos** de las variables del conjunto de datos. El INEGI codifica las variables con números o factores, los cuales corresponden a valores. Por ejemplo, para la variable de `sexo` 1 corresponde a Hombre, 2 a Mujer y 9 a No específicado.
- `modelo_entidad_relacion` Ilustra la estructura del conjunto de datos.


### Diccionario de datos

Un diccionario de datos proporciona una descripción detallada de las variables contenidas en un conjunto de datos, especificando su significado, tipo, formato y cualquier otra información relevante para su correcta interpretación y uso. El objetivo del diccionario de datos es garantizar la consistencia, claridad y estándares comunes entre todos los usuarios de los datos, facilitando su comprensión y reduciendo errores.

Un diccionario de datos debe contener.
* **Variable**: Nombre de la variable tal como aparece en el conjunto de datos.
* **Tipo**: Clasificación del tipo de dato (por ejemplo, numérico, texto, fecha, booleano, etc.).
* **Descripción**: Explicación detallada sobre el significado de la variable y cómo debe interpretarse.

Además, es útil que incluya:
* **Ejemplo**: Un valor de ejemplo representativo para ilustrar el formato y la información de la variable.
* **Formato**: Especificación del formato esperado para la variable (por ejemplo, "dd/mm/aaaa" para fechas o números con dos decimales).
* **Notas y aclaraciones**: Cualquier nota adicional o consideración especial que sea importante para la interpretación correcta de la variable.

Los estándares internacionales son normas, reglas o directrices establecidas por organismos reconocidos a nivel mundial para garantizar la interoperabilidad, calidad, seguridad y uniformidad en distintos ámbitos, como la tecnología, la industria, la salud y los datos. Estos estándares facilitan la comunicación y el intercambio de información entre sistemas, instituciones y países, asegurando que todos trabajen bajo las mismas reglas. Seguir estos estándares al estructurar los datos facilita la integración de datos provenientes de diferentes fuentes y se mejora la calidad de la información.

Algunos ejemplos de estándares internacionales:
* [ISO](www.iso.org) (International Organization for Standardization): ISO 11179 para metadatos y diccionarios de datos
* [W3C](www.w3.org) (World Wide Web Consortium): DCAT para catalogación de datos abiertos incluyendo datos de gobierno
* [LOINC](loinc.org) y [SNOMED](www.snomed.org): ICE-10 u SNOMED CT para terminología médica
* [DCMI](www.dublincore.org) (Dublin Core Metadata Initiative): Dublin Core para descripción de recursos digitales

Es importante recalcar las diferencias entre diccionario de datos, metadata y diccionario de Python:
* Diccionario de Datos (de un conjunto de datos). Es un documento estructurado que describe las variables de un conjunto de datos. Incluye nombre de la variable, tipo de dato, descripción.
* Metadatos. Son datos sobre los datos, proporcionando información adicional sobre un archivo, conjunto de datos o sistema. Incluyen detalles como fuente, fecha de creación, propietario, permisos de acceso y estructura.
* Diccionario en Python (`dict`). Es una estructura de datos en Python que almacena pares clave-valor. Por ejemplo: `persona = {"nombre": "Ana", "edad": 25, "pais": "Mexico"}`


En el caso del EDR podemos consultar el diccionario de datos incluido en `diccionario_datos_defunciones_registradas_2020.csv`, el cual forma parte del conjunto de datos. El incluir el diccionario de datos permite entender la información incluida, el contexto de los datos, cómo fueron generados y sus limitaciones.

### Ejercicio: revisar el diccionario de datos

Tómate un momento ahora para abrir los archivos `metadatos_defunciones_registradas_2020.txt` y `diccionario_datos_defunciones_registradas_2020.csv`.

Revisa el índice o la tabla de contenido y fíjate en la estructura del documento. Busca las secciones que describen las variables del conjunto de datos principal y los catálogos. No necesitas memorizar todo, pero es crucial que sepas dónde encontrar esta información cuando la necesites para interpretar los datos más adelante.

Responde las siguientes preguntas:

1. ¿Qué información esperas encontrar en el conjunto de datos?
2. ¿Qué información está contenida en cada archivo?
    * `metadatos_defunciones_registradas_2020.txt`
    * `diccionario_datos_defunciones_registradas_2020.csv`
    * `conjunto_de_datos_defunciones_registrados_2020.csv`
3. ¿En que archivo se encuentra el catálogo de `edad`? ¿Cómo se codifica esta variable en el conjunto de datos?
4. ¿Cuáles son las limitaciones del EDR?

Ahora que ya conoces tus archivos, el siguiente paso es cargarlos para empezar a trabajar con Python. Vimos como hacer esto en la sección [Colab: uso, archivos y bibliotecas](https://drive.google.com/file/d/1jHpCfFSxAwFHpT5pqZqHvalOcO_X1IIp/view?usp=sharing).

Una vez que los datos han sido cargados en un DataFrame, es fundamental realizar una verificación rápida. Este paso te permite confirmar que el proceso de carga fue exitoso y que los datos se almacenaron con la estructura y los tipos de datos correctos.

> Nota: La versión gratuita de Google Colab tiene recursos limitados, por lo que usaremos solo las primeras 10000 filas.


In [42]:
# Instalar e importar bibliotecas
import pandas as pd

# Abrir datos
from google.colab import drive
drive.mount('/content/drive')

file_name = '/content/drive/MyDrive/Trabajo/Python_EDR2020/conjunto_de_datos_defunciones_registradas_2020_csv/conjunto_de_datos/conjunto_de_datos_defunciones_registrados_2020.csv'
df = pd.read_csv(file_name, nrows=10000)
df


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
0,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,9,88,88,888,8888,0,999
1,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
2,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
3,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
4,1,6,1,9,4,1,1,9,4,1,...,2020,NaN,2,1,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,2,2,2,2,3,211,2,2,3,211,...,2020,NaN,9,9,88,88,888,8888,0,999
9996,2,2,2,2,4,227,2,2,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9997,2,2,2,2,1,2135,2,2,4,268,...,2020,NaN,2,2,88,88,888,8888,0,999
9998,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999


## Conjuntos de datos ordenados (Tidy Data)


Para el análisis de datos debemos de familiarizarnos con varias definiciones, usando como ejemplo el EDR:

* **Variable**: Un atributo que estamos midiendo. En la tabla de defunciones, el **sexo**, la **edad** y la **causa de muerte** son variables.
* **Valor**: La medida específica de una variable para una observación. Si la variable es "sexo", el valor podría ser "Hombre". Si la variable es "edad", el valor podría ser "25".
* **Observación**: El conjunto de todos los valores medidos para una unidad de estudio. En nuestra tabla, **una fila entera** es una observación, ya que contiene todos los valores (sexo, edad, causa, etc.) para una defunción específica.


Para trabajar de manera efectiva con los datos, es crucial entender los conceptos detrás de su organización. Hadley Wickham, una figura clave en la ciencia de datos, propuso los principios de los **Datos Ordenados (Tidy Data)**. Un conjunto de datos ordenado es aquel que está estructurado de una forma estándar y predecible, lo que facilita enormemente su limpieza y análisis.

Un conjunto de datos se considera ordenado cuando cumple con tres reglas fundamentales:
1.  **Cada variable forma una columna.**  Esto significa que cada característica que queremos analizar (edad, sexo, causa de muerte) tiene su propia columna.
2.  **Cada observación forma una fila.** Cada registro individual (en nuestro caso, cada defunción) es una fila única.
3.  **Cada tipo de unidad de observación forma una tabla.** Si tuviéramos datos de defunciones y de nacimientos, cada uno debería ir en una tabla separada.

El equivalente de "Tidy Data" en el mundo de las bases de datos relacionales es la normalización de bases de datos.

Aunque nuestro conjunto de datos de las EDR ya está bien estructurado, no siempre es así. Los datos que recibimos inicialmente, a menudo llamados datos crudos, suelen estar "desordenados" (untidy). Podrían tener varias variables en una sola columna o información repetida. En muchos proyectos, la limpieza de datos incluye el transformar esos datos crudos en un conjunto de datos ordenado y listo para ser analizado, siguiendo los principios que acabamos de aprender.




## Tipos de datos

Los datos no son todos iguales. Es importante identificar su tipo, ya que esto determina qué operaciones podemos realizar con ellos.

* **Numéricos** (`int` o `float`): Valores que se pueden contar o medir. Pueden ser discretos (números enteros, como la cantidad de hijos) o continuos (números decimales, como la edad).
* **Categóricos** (`category`): Valores que pertenecen a un conjunto limitado de categorías. Por ejemplo, el **sexo** (hombre/mujer) o el **estado civil** (soltero, casado, etc.).
* **Binarios** (`bool`): Un tipo de dato categórico con solo dos valores posibles, como "Verdadero" o "Falso".
* **Fecha** (`datetime64`): Representan un punto en el tiempo.
* **Texto** (`object`): Cadenas de caracteres que no se usan para cálculos.
* **Coordenadas**: Datos de ubicación geográfica.


Es común que los datos estén incompletos. En `pandas`, los **valores faltantes** se representan con `NaN` (Not a Number) o `None`. Este es un tema muy complejo al que regresaremos más adelante.

Cuando cargas un archivo (como un CSV o un DBF), este solo contiene texto. Pandas no "sabe" de antemano qué tipo de dato es cada columna. Su trabajo es leer el archivo y adivinar inteligentemente el tipo de dato basándose en el contenido de cada columna. Por ejemplo, si una columna solo contiene números, Pandas la inferirá como un tipo numérico (`int` o `float`). Si una columna contiene texto, fechas o una mezcla de tipos, la inferirá como `object`. Esta inferencia automática te ahorra mucho tiempo, pero a veces puede ser inexacta (por ejemplo, al inferir números de identificación como valores numéricos en lugar de texto), por lo que es vital revisarla con `df.info()`.

**Nota sobre `object`**. Pandas asigna el tipo `object` cuando una columna contiene:
* Cadenas de texto `strings`.
* Una mezcla de tipos de datos, como números y texto.
* Valores faltantes (`None` o `NaN`) en una columna que originalmente era de números enteros, ya que el tipo `int` no puede representar valores nulos.
Piensa en él como el tipo de dato que almacena cualquier cosa que no sea un número o una fecha clara, como nombres, descripciones, identificadores, etc. Es importante estar atento a las columnas de tipo `object` porque, aunque parezcan contener números, Pandas no puede realizar operaciones matemáticas directamente sobre ellas. En un análisis de datos, a menudo necesitarás limpiar y convertir estas columnas al tipo de dato correcto antes de usarlas.

### Ejercicio: datos ordenados

Ahora, con las definiciones claras, aplica estos conceptos a tu conjunto de datos de las EDR.
Identifica en la tabla que estás trabajando:
1.  **Variables, valores y observaciones**: ¿Qué representa cada una de estas definiciones en la tabla?
2.  **Tipos de datos**: ¿Qué tipos de datos podrías encontrar en las columnas del DataFrame de las EDR? (e.g., ¿la edad es numérica?, ¿el género es categórico?).

In [ ]:
df.info()

## Estructura de un DataFrame

Un **DataFrame** de **pandas**, la estructura principal con la que trabajaremos, es similar a una hoja de cálculo o una tabla de una base de datos. Está compuesto por tres elementos esenciales: los **valores**, las **columnas** y el **índice**.

  * **Valores**: Son los datos que contiene la tabla. Puedes acceder a ellos usando `.values`, que los devuelve como una matriz de NumPy.
  * **Columnas**: Son los nombres de cada variable. Puedes verlos como la fila superior de la tabla. Para ver solo los nombres de las columnas, usa el atributo `.columns`. Es un atributo, no una función, por lo que no lleva paréntesis.
  * **Índice**: Es la etiqueta de cada fila. Por defecto, pandas asigna un índice numérico secuencial (0, 1, 2, ...), pero puede ser personalizado. Puedes ver el índice con el atributo `.index`.

Además de estos elementos, hay otros atributos muy útiles para entender la estructura de tu DataFrame:

  * **`.shape`**: Te muestra el número total de filas y columnas del DataFrame. Por ejemplo, `(20000, 50)` significa que hay 20,000 filas y 50 columnas.
  * **`.dtypes`**: Te devuelve una lista de los tipos de datos de cada columna, lo que te ayuda a verificar si pandas infirió los tipos de datos correctamente durante la carga.

Para ver los nombres de las columnas, simplemente ejecuta: `df.columns`.

Nota: `.columns`, `.index`, `.values`, `.shape` y `.dtypes` son **atributos** del DataFrame, no métodos. Por ello, se escriben sin los paréntesis `()`.

In [41]:
df.columns

Index(['ent_regis', 'mun_regis', 'ent_resid', 'mun_resid', 'tloc_resid',
       'loc_resid', 'ent_ocurr', 'mun_ocurr', 'tloc_ocurr', 'loc_ocurr',
       'causa_def', 'lista_mex', 'sexo', 'edad', 'dia_ocurr', 'mes_ocurr',
       'anio_ocur', 'dia_regis', 'mes_regis', 'anio_regis', 'dia_nacim',
       'mes_nacim', 'anio_nacim', 'ocupacion', 'escolarida', 'edo_civil',
       'presunto', 'ocurr_trab', 'lugar_ocur', 'necropsia', 'asist_medi',
       'sitio_ocur', 'cond_cert', 'nacionalid', 'derechohab', 'embarazo',
       'rel_emba', 'horas', 'minutos', 'capitulo', 'grupo', 'lista1',
       'gr_lismex', 'vio_fami', 'area_ur', 'edad_agru', 'complicaro',
       'dia_cert', 'mes_cert', 'anio_cert', 'maternas', 'lengua', 'cond_act',
       'par_agre', 'ent_ocules', 'mun_ocules', 'loc_ocules', 'razon_m',
       'dis_re_oax'],
      dtype='object')

La primera columna que observas en la tabla, la que no tiene nombre, es el **índice** del DataFrame. Por defecto, pandas asigna a cada fila un índice numérico que comienza en cero y va en aumento.

El índice es una etiqueta única para cada fila, no es una columna de datos. Se utiliza para identificar y acceder a las filas de manera eficiente. Si no especificamos una columna para que sirva como índice al cargar los datos, pandas simplemente los enumera de forma automática.

Para ver el índice de tu DataFrame, usa el atributo `.index`:

In [ ]:
df.index

## Selección (slicing)

Trabajar con la tabla completa no siempre es necesario. A menudo, necesitamos seleccionar solo una porción de los datos, ya sean columnas, filas o un subconjunto específico. A este proceso se le conoce como **slicing**. Pandas ofrece varias formas de hacerlo.

Idealmente, los nombres de columnas y del índice  deben de ser únicos, para que sea posible referirse sin ambigüedad a cada celda.

Para seleccionar una **única columna**, usa corchetes (`[]`) y el nombre de la columna como una cadena de texto. El resultado es una **`pandas.Series`**.

In [43]:
df['sexo']

,sexo
0,2
1,2
2,2
3,1
4,1
...,...
9995,2
9996,1
9997,2
9998,1


Para seleccionar **varias columnas** a la vez, pasa una **lista** de nombres de columnas dentro de un par de corchetes. El resultado es un nuevo **DataFrame**. Por esta razón, se le conoce como el uso de "doble corchete".

In [44]:
df[ ['sexo','edad'] ]

,sexo,edad
0,2,4074
1,2,4087
2,2,4062
3,1,4080
4,1,4078
...,...,...
9995,2,4091
9996,1,4054
9997,2,4094
9998,1,4047


El método `.loc` se usa para seleccionar filas y columnas basándose en sus **nombres** o **etiquetas**. Se accede a él de la misma manera que a un diccionario. `df.loc[filas, columnas]`

Puedes seleccionar una fila específica por su nombre de índice, o un rango de filas de la misma forma que lo harías en una lista. También puedes pasar una lista para seleccionar filas no consecutivas.

In [45]:
df.loc[10]

,10
ent_regis,1
mun_regis,1
ent_resid,1
mun_resid,3
tloc_resid,7
loc_resid,1
ent_ocurr,1
mun_ocurr,3
tloc_ocurr,7
loc_ocurr,1


También puedes seleccionar un subconjunto específico de filas y columnas al mismo tiempo.

In [46]:
df.loc[ 7:13 , ['sexo','edad'] ]

,sexo,edad
7,2,4075
8,1,4073
9,2,4066
10,2,4063
11,2,4081
12,2,4075
13,1,4069


El método `.iloc` (con la "i" de *integer*) se utiliza para seleccionar filas y columnas basándose en su **posición numérica**, similar a como se indexan los elementos en una lista de Python.

Usa la misma lógica para seleccionar un subconjunto por posición, solo que en lugar de etiquetas utiliza posiciones. En este caso para el índice las etiquetas son las posiciones, pero puedes notar la diferencia en las columnas.

In [47]:
df.iloc[7:13, [15, 23]]

,mes_ocurr,ocupacion
7,2,11
8,2,99
9,7,4
10,4,11
11,12,11
12,12,11


En resumen, utiliza `.loc` cuando trabajes con etiquetas (nombres de columnas o índice personalizado), y `.iloc` cuando trabajes con posiciones numéricas.

## Ordenar datos

Para cambiar el orden de las columnas de un DataFrame, la forma más sencilla es crear una lista con los nombres de las columnas en el orden deseado y usarla para indexar tu DataFrame.

Por ejemplo, imagina que tu DataFrame `df` tiene muchas columnas, pero solo te interesan `'sexo'`, `'edad'` y `'causa_def'`. Además, quieres que `'causa_def'` sea la primera columna. Simplemente, crea una lista con ese orden y aplícala:

In [48]:
df[['causa_def', 'sexo', 'edad']]

,causa_def,sexo,edad
0,J441,2,4074
1,K566,2,4087
2,I120,2,4062
3,J869,1,4080
4,I709,1,4078
...,...,...,...
9995,J439,2,4091
9996,U071,1,4054
9997,I219,2,4094
9998,U071,1,4047


Es muy importante entender que esta operación **no modifica el DataFrame original**. Lo que hace es devolver una **vista temporal** del DataFrame con el nuevo orden. Para que los cambios persistan, es necesario **guardar** este resultado en una nueva variable o reasignarla a la variable original.

Este principio de reasignación es fundamental en pandas, ya que la mayoría de las operaciones de modificación no actúan sobre el objeto original por defecto.

In [49]:
df_reordenado = df[['causa_def', 'sexo', 'edad']]
df_reordenado

,causa_def,sexo,edad
0,J441,2,4074
1,K566,2,4087
2,I120,2,4062
3,J869,1,4080
4,I709,1,4078
...,...,...,...
9995,J439,2,4091
9996,U071,1,4054
9997,I219,2,4094
9998,U071,1,4047


La función **`.sort_values()`** de pandas te permite ordenar un DataFrame o una Serie basándote en los valores de una o más columnas. Ten en cuenta que:
* Por defecto, pandas ordena de forma ascendente (de menor a mayor o alfabéticamente). Para ordenar de forma descendente, usa el parámetro `ascending=False`.
* `NaN` (valores faltantes) siempre se colocan al final de la tabla por defecto.
* En datos de texto, los caracteres especiales y los acentos pueden afectar el orden alfabético.

Se puede notar en nuevo orden en el índice.

In [50]:
df_reordenado.sort_values(by='edad')

,causa_def,sexo,edad
9488,P280,1,1001
427,P072,1,1001
9400,P072,1,1001
9475,P220,1,1001
8955,P072,1,1001
...,...,...,...
4932,V093,1,4998
8786,V093,1,4998
5885,X704,1,4998
6908,X708,1,4998


Puedes ordenar los datos por varias columnas a la vez. Para ello, pasa una **lista** de nombres de columnas al parámetro `by`. También puedes especificar una dirección de ordenación diferente para cada columna usando una lista en el parámetro `ascending`.

Es importante entender la sintaxis de pandas:

1.  Selección de datos: primero, eliges sobre qué vas a trabajar (`df` o `df['columna']`).
2.  Llamado a la función: luego, llamas a la función que quieres usar (`.sort_values()`).
3.  Parámetros: finalmente, pasas las instrucciones específicas para la función dentro de los paréntesis (`by='edad'`, `ascending=False`, etc.).

In [51]:
df_reordenado.sort_values(by=['sexo', 'edad'], ascending=[True, False])

,causa_def,sexo,edad
630,Y349,1,4998
1790,V093,1,4998
3027,Y244,1,4998
3118,U071,1,4998
4171,Y099,1,4998
...,...,...,...
9491,P073,2,1002
9951,P220,2,1002
6694,Y349,9,4998
9594,U071,9,3002


**Nota**: Cuando ordenas por una sola columna y hay filas con valores idénticos, el orden de estas filas no está garantizado (es arbitrario). Si necesitas un orden consistente, es una buena práctica ordenar por una segunda columna para "romper" esos empates. Por ejemplo, `df.sort_values(by=['edad', 'sexo'])` garantizará que si dos personas tienen la misma edad, el orden entre ellas se determine por su sexo.


Reordenar el índice de un DataFrame es una operación diferente a reordenar las columnas, ya que el índice no es un dato de la tabla, sino una etiqueta.

El método `.reindex()` te permite reordenar el índice de tu DataFrame. Por defecto, si un índice no existe en el nuevo orden, se llenará con valores `NaN`.

El método `.sort_index()` también te permite ordenar el índice de forma alfabética o numérica, de manera similar a como `.sort_values()` ordena las columnas.

## Eliminar filas y columnas

La función `.drop()` permite eliminar filas o columnas de un DataFrame. Es importante recordar que, por defecto, `.drop()` devuelve una nueva copia del DataFrame sin las filas o columnas eliminadas. Para que los cambios persistan en tu DataFrame original, debes guardar el resultado o usar el parámetro `inplace=True`.

Para eliminar una o más filas, le pasas las etiquetas del índice a .drop(). Puedes pasar un solo índice o una lista de índices.

In [52]:
df.drop(index=[0, 1, 2])

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
3,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
4,1,6,1,9,4,1,1,9,4,1,...,2020,NaN,2,1,88,88,888,8888,0,999
5,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
6,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
7,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,2,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,2,2,2,2,3,211,2,2,3,211,...,2020,NaN,9,9,88,88,888,8888,0,999
9996,2,2,2,2,4,227,2,2,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9997,2,2,2,2,1,2135,2,2,4,268,...,2020,NaN,2,2,88,88,888,8888,0,999
9998,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999


Para eliminar columnas, debes usar el parámetro `columns` dentro de la función `.drop()`. Al igual que con las filas, puedes pasarle un solo nombre de columna o una lista de nombres de columnas.

También puedes usar el parámetro `axis=1` para eliminar columnas, este le indica a pandas que la operación se debe realizar sobre las columnas (eje 1), y no sobre las filas (eje 0, que es el valor por defecto). Sin embargo, la forma más clara y recomendada es usar el parámetro `columns`.

Si quieres modificar el DataFrame original sin crear una nueva copia, puedes usar el parámetro `inplace=True`. Ten en cuenta que, una vez que usas este parámetro, los cambios son permanentes.

In [53]:
df.drop(columns=['ent_regis','mun_regis','tloc_resid'])

,ent_resid,mun_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,causa_def,lista_mex,sexo,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
0,1,1,1,1,1,15,1,J441,33G,2,...,2020,NaN,2,9,88,88,888,8888,0,999
1,1,1,1,1,1,15,1,K566,35I,2,...,2020,NaN,9,9,88,88,888,8888,0,999
2,1,1,1,1,1,15,1,I120,27Z,2,...,2020,NaN,2,2,88,88,888,8888,0,999
3,1,1,1,1,1,15,1,J869,33Z,1,...,2020,NaN,9,9,88,88,888,8888,0,999
4,1,9,1,1,9,4,1,I709,31A,1,...,2020,NaN,2,1,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,2,2,211,2,2,3,211,J439,33D,2,...,2020,NaN,9,9,88,88,888,8888,0,999
9996,2,2,227,2,2,15,1,U071,06T,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9997,2,2,2135,2,2,4,268,I219,28A,2,...,2020,NaN,2,2,88,88,888,8888,0,999
9998,2,1,1,2,1,14,1,U071,06T,1,...,2020,NaN,2,1,88,88,888,8888,0,999


## Filtrado de datos

Este proceso de filtrado nos permite trabajar con subconjuntos de datos de interés, seleccionando solo las filas que cumplen con ciertas condiciones, lo que es ayuda para responder preguntas específicas. En pandas, esto se conoce como **filtrado booleano**.

Nota: Al realizar los ejemplos, notarás que los valores de las columnas `causa_def` son códigos numéricos. Para entender qué significa cada uno de estos códigos, es necesario consultar el catálogo de defunciones que viene en los archivos que descargaste del INEGI. Este es un ejemplo de por qué la exploración de metadatos es un paso importante en el análisis de datos.

Para seleccionar las filas que cumplen con una condición, debes:

1.  Crear una **condición booleana** que devuelva `True` o `False` para cada fila.
2.  Pasar esa condición al DataFrame dentro de corchetes.

Por ejemplo, para seleccionar las filas donde el sexo es femenino (según el catálogo, el valor es `2`) usamos un doble igual `==`:

In [54]:
df['sexo'] == 2

,sexo
0,True
1,True
2,True
3,False
4,False
...,...
9995,True
9996,False
9997,True
9998,False


Usa la condición para filtrar el DataFrame. Aunque usar `df[condicion]` es la forma más común y sencilla de filtrar, es buena práctica usar `.loc`.

La ventaja de `.loc` es que hace explícito que estás seleccionando filas basándote en su etiqueta (en este caso, una etiqueta booleana), lo que puede hacer tu código más legible. Además, es la forma más segura de evitar errores futuros al combinar la selección de filas y columnas.

In [56]:
df.loc[ df['sexo'] == 2 ]

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
3,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
4,1,6,1,9,4,1,1,9,4,1,...,2020,NaN,2,1,88,88,888,8888,0,999
5,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
8,1,1,1,6,8,1,1,6,8,1,...,2020,NaN,2,9,88,99,999,9999,0,999
13,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9990,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9991,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9996,2,2,2,2,4,227,2,2,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9998,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999


De forma similar, puedes usar operadores como "diferente de" (`!=`) para seleccionar las filas que no cumplen la condición.

In [57]:
df.loc[df['sexo'] != 2]

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
3,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
4,1,6,1,9,4,1,1,9,4,1,...,2020,NaN,2,1,88,88,888,8888,0,999
5,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
8,1,1,1,6,8,1,1,6,8,1,...,2020,NaN,2,9,88,99,999,9999,0,999
13,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9990,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9991,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9996,2,2,2,2,4,227,2,2,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9998,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999


Si los datos son numéricos, puedes usar operadores de comparación como "mayor que" (`>`), "menor que" (`<`), "mayor o igual que" (`>=`), "menor o igual que" (`<=`).

Por ejemplo, para seleccionar las filas de decesos de personas de más de 65 años (según el catálogo, el valor es `4065`):

In [58]:
df.loc[ df['edad'] > 4065 ]

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
0,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,9,88,88,888,8888,0,999
1,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
3,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
4,1,6,1,9,4,1,1,9,4,1,...,2020,NaN,2,1,88,88,888,8888,0,999
7,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,2,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9988,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9991,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9992,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,2,88,88,888,8888,0,999
9995,2,2,2,2,3,211,2,2,3,211,...,2020,NaN,9,9,88,88,888,8888,0,999


Los datos faltantes (`NaN`) son un problema común. Para identificar las filas que tienen valores nulos en una columna específica, se usa el método `.isna()`.

Nota: En el caso de los datos del INEGI no hay faltantes con `NaN`, ya que los registros sin datos se codifican de otras maneras.

In [59]:
df.loc[ df['maternas'].isna() ]

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
0,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,9,88,88,888,8888,0,999
1,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
2,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
3,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
4,1,6,1,9,4,1,1,9,4,1,...,2020,NaN,2,1,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,2,2,2,2,3,211,2,2,3,211,...,2020,NaN,9,9,88,88,888,8888,0,999
9996,2,2,2,2,4,227,2,2,15,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9997,2,2,2,2,1,2135,2,2,4,268,...,2020,NaN,2,2,88,88,888,8888,0,999
9998,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,1,88,88,888,8888,0,999


Cuando necesitas seleccionar filas que coincidan con varios valores específicos, es más eficiente usar el método `.isin()`. Este método toma una lista de valores y devuelve `True` si el valor de la columna está en esa lista.

Por ejemplo, cinco causas más comunes de defunción (usando los datos completos) son: 'I219', 'E116', 'J189', 'E117', 'E112'. Usemos esa lista para seleccionar los registros asociados a esas causas.

In [60]:
df.loc[  df['causa_def'].isin( ['I219', 'E116', 'J189', 'E117', 'E112'] )  ]

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
7,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,2,88,88,888,8888,0,999
9,1,6,1,6,8,1,1,6,8,1,...,2020,NaN,2,1,88,88,888,8888,0,999
28,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
34,1,2,1,2,2,7,1,2,2,7,...,2020,NaN,9,2,88,88,888,8888,0,999
54,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9970,2,2,2,2,2,143,2,2,2,143,...,2020,NaN,2,2,88,88,888,8888,0,999
9972,2,2,2,2,4,456,2,2,4,456,...,2020,NaN,2,2,88,88,888,8888,0,999
9984,2,4,2,4,16,1,99,999,99,9999,...,2020,NaN,2,2,88,88,888,8888,0,999
9985,2,4,2,4,16,1,2,4,16,1,...,2020,NaN,2,2,88,88,888,8888,0,999


Para combinar varias condiciones de filtrado, debes encerrar cada condición entre paréntesis y unirlas con operadores lógicos:

  * `&`: Para combinar condiciones con "y" (`AND`). Se deben cumplir ambas.
  * `|`: Para combinar condiciones con "o" (`OR`). Se debe cumplir al menos una.
  * `~`: Para negar una condición.

En selecciones complejas, 3s una buena práctica realizar las selecciones por pasos y guardar los resultados en variables con nombres descriptivos.

Para seleccionar los registros de mujeres menores de un año, se selecciona las filas donde el sexo es 2 (femenino) **Y** la edad es menor de 4001 (un año).

In [61]:
df.loc[ (df['sexo'] == 2) & (df['edad'] < 4001) ]

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
55,1,3,1,3,1,94,1,1,15,1,...,2020,NaN,8,8,88,88,888,8888,0,999
344,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,8,8,88,88,888,8888,0,999
440,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,8,8,88,88,888,8888,0,999
1032,1,1,1,11,7,1,1,1,15,1,...,2020,NaN,8,8,88,88,888,8888,0,999
1129,1,3,1,3,7,1,1,3,7,1,...,2020,NaN,8,8,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9502,1,5,1,1,15,1,1,5,1,469,...,2020,NaN,8,8,88,88,888,8888,0,999
9505,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,8,8,88,88,888,8888,0,999
9510,1,11,1,11,7,1,1,11,7,1,...,2020,NaN,8,8,88,88,888,8888,0,999
9904,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,8,8,88,88,888,8888,0,999


Las columnas tienen funciones especializadas dependiendo de su tipo. Por ejemplo, las columnas de texto permiten hacer operaciones de `strings`.

Para seleccionar todas las defunciones por enfermedades respiratorias, que en el catálogo del CIE-10 son códigos que inician con la letra `J`, puedes usar el método `.str.startswith()`:

In [62]:
df.loc[ df['causa_def'].str.startswith('J') ]

,ent_regis,mun_regis,ent_resid,mun_resid,tloc_resid,loc_resid,ent_ocurr,mun_ocurr,tloc_ocurr,loc_ocurr,...,anio_cert,maternas,lengua,cond_act,par_agre,ent_ocules,mun_ocules,loc_ocules,razon_m,dis_re_oax
0,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,9,88,88,888,8888,0,999
3,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
17,1,1,1,3,7,1,1,3,7,1,...,2020,NaN,2,1,88,88,888,8888,0,999
21,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
32,1,1,1,1,15,1,1,1,15,1,...,2020,NaN,2,2,88,88,888,8888,0,999
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9597,1,7,1,7,8,1,1,7,8,1,...,2020,NaN,8,8,88,88,888,8888,0,999
9937,2,4,2,5,11,1,2,4,16,1,...,2020,NaN,2,1,88,88,888,8888,0,999
9976,2,2,2,2,15,1,2,2,15,1,...,2020,NaN,9,9,88,88,888,8888,0,999
9989,2,1,2,1,14,1,2,1,14,1,...,2020,NaN,2,2,88,88,888,8888,0,999


## Conteo y valores únicos

Para conocer los valores que existen en una columna, puedes usar la función **`.unique()`**. Esto te devuelve una lista (un arreglo de NumPy, para ser exactos) de todos los valores distintos que se encuentran en esa columna. Por ejemplo, si revisamos el perfil de datos de las EDR, veremos que hay múltiples causas de defunción. Usando `.unique()`, podemos obtener una lista de todas ellas.

In [63]:
df['causa_def'].unique()

array(['J441', 'K566', 'I120', 'J869', 'I709', 'U071', 'E149', 'E117',
       'X599', 'I64X', 'C509', 'I509', 'E870', 'C910', 'G931', 'J841',
       'C61X', 'K746', 'J180', 'X590', 'I110', 'E46X', 'N390', 'F142',
       'X700', 'E43X', 'J449', 'I219', 'N189', 'D430', 'I619', 'K650',
       'C530', 'D489', 'I678', 'K255', 'I609', 'C189', 'N185', 'N179',
       'K260', 'V496', 'N498', 'E112', 'J188', 'V439', 'X740', 'E116',
       'Q913', 'K801', 'I00X', 'J189', 'B209', 'K830', 'K610', 'F102',
       'A090', 'I501', 'I613', 'E668', 'J448', 'N40X', 'J159', 'X708',
       'D649', 'C349', 'I249', 'G122', 'A419', 'E785', 'J440', 'M321',
       'I499', 'I139', 'I420', 'C679', 'K729', 'N281', 'C829', 'C719',
       'C229', 'E872', 'G409', 'Q639', 'J209', 'I872', 'K743', 'K550',
       'B159', 'N200', 'I629', 'C169', 'J961', 'G301', 'I10X', 'D469',
       'M329', 'C809', 'G309', 'C259', 'G10X', 'V289', 'N009', 'C419',
       'E119', 'G20X', 'K920', 'K750', 'K922', 'X470', 'L100', 'P369',
      

Si solo te interesa saber cuántos valores únicos hay, y no cuáles son, la función **`.nunique()`** devuelve un número entero con el conteo.

In [64]:
df['causa_def'].nunique()

907

La función **`.value_counts()`** es una de las más utilizadas en la exploración de datos. Te permite contabilizar la frecuencia con la que aparece cada valor único en una columna. El resultado es una `pd.Series` que tiene los valores únicos como índice y su respectiva cuenta como valor.

In [68]:
df['causa_def'].value_counts()

,count
causa_def,
U071,2053
I219,796
U072,437
J189,335
E112,195
...,...
A414,1
L038,1
B212,1


Aquí, `causa_def` es el nombre del índice, y los códigos son sus etiquetas. El valor numérico a la derecha de cada código es su frecuencia. Por ejemplo, la primera fila es `U071   2053`, lo que significa que para el código `U071` (COVID-19) hubo 2053 defunciones en nuestros datos.
El nombre `count` que se muestra en la esquina superior derecha es el nombre de la columna de valores generada por la función.
A la mitad de la salida hay tres puntos `...`, ya que por el tamaño no se despliega toda la tabla, sino solo los primeros y últimos 5 valores.
La información al final `907 rows × 1 columns, dtype: int64` te dice que hay 907 códigos de causa de defunción únicos y que el conteo está almacenado como un número entero de 64 bits.

La función `value_counts()` ignora los valores faltantes (`NaN`) por defecto. Para incluirlos en el conteo, debes usar el parámetro `dropna=False`. Esto es crucial al calcular porcentajes, ya que el denominador cambiará.
Si lo que necesitas es ver la **distribución en porcentajes** en lugar de conteos absolutos, puedes usar el parámetro `normalize=True`.

En el caso de los datos del INEGI, los faltantes se codifican con una clave, por ejemplo `9999` qué pandas trata cómo un valor. Sin embargo, recuerda esto para otras tablas.

In [69]:
df['causa_def'].value_counts(normalize=True, dropna=False)

,proportion
causa_def,
U071,0.2053
I219,0.0796
U072,0.0437
J189,0.0335
E112,0.0195
...,...
A414,0.0001
L038,0.0001
B212,0.0001


También puedes aplicar `.value_counts()` a **varias columnas** al mismo tiempo, lo que te permite ver la frecuencia de combinaciones de valores.

Cuando aplicas `.value_counts()` a varias columnas, como `[['SEXO', 'EDAD']]`, Pandas agrupa las filas por la combinación única de valores en esas columnas. El resultado es una `pd.Series` cuyo índice se convierte en un índice jerárquico o `MultiIndex`.

In [72]:
df[['sexo', 'edad']].value_counts()

sexo  edad
1     4074    136
      4065    134
      4067    133
      4070    129
      4075    128
             ... 
      2006      1
      2011      1
      2010      1
      2026      1
      2025      1
Name: count, Length: 288, dtype: int64

La función `.reset_index()` permite transformar la `pd.Series` con `MultiIndex` en un DataFrame más fácil de usar. Al aplicarla, los niveles del `MultiIndex` se convierten en columnas regulares del nuevo DataFrame. La columna de conteo, que por defecto se llama `count`, también se vuelve una columna más.

Este proceso también se puede usar para transformar series a DataFrame.

In [73]:
df[['sexo', 'edad']].value_counts().reset_index()

,sexo,edad,count
0,1,4074,136
1,1,4065,134
2,1,4067,133
3,1,4070,129
4,1,4075,128
...,...,...,...
283,1,2006,1
284,1,2011,1
285,1,2010,1
286,1,2026,1


Dado que el resultado de **`.value_counts()`** es una `pd.Series` (o un DataFrame si usas `.reset_index()`, puedes usar otras funciones de pandas, como `.head()`, para ver los resultados más importantes. Por ejemplo, para ver las cinco causas de defunción más comunes:

In [ ]:
df['causa_def'].value_counts().head(5)

## Selección y revisión de un subconjunto

El EDR tiene una gran cantidad de datos, por los que muchas veces seleccionamos solo las variables y observaciones que ayudan a responder nuestra pregunta de interes. Por ejemplo, si trabajamos con las enfermedades del sistema respiratorio podemos identificar las siguientes variables del conjunto de datos. En la etapa de limpieza, nos centraremos en estas columnas y eliminaremos el resto para optimizar el análisis.

| Categoría | Variables del INEGI |
| :--- | :--- |
| **Tiempo** | `'DIA_OCURR'`, `'MES_OCURR'`, `'ANIO_OCUR'` |
| **Lugar** | `'ENT_OCURR'`, `'MUN_OCURR'`, `'AREA_UR'` |
| **Causas defunción** | `'CAUSA_DEF'` |
| **Grupo etario** | `'SEXO'`, `'EDAD'`, `'EDAD_AGRU'` |

Para seleccionar el subconjunto de datos que nos interesa, seguiremos dos pasos:
1. **Filtrar por variables**: Primero, seleccionaremos únicamente las columnas de interés que definimos en la tabla anterior.
2. **Filtrar por observaciones**: Luego, seleccionaremos solo las filas que corresponden a defunciones de interés. Por ejemplo, si nos interesan las enfermedades del sistema respiratorio podemos usar la columna `'CAUSA_DEF'` para seleccionar aquellas con códigos que comienzan con la letra `'J'`.

Nota como guardamos el resultado de la selección en un nuevo DataFrame llamado `df_respiratorio`.


In [74]:
df.shape

(10000, 59)

In [75]:
# Paso 1: Seleccionar las columnas de interés
columnas = ['dia_ocurr', 'mes_ocurr', 'anio_ocur',
            'ent_ocurr', 'mun_ocurr', 'causa_def',
            'sexo', 'edad', 'edad_agru', ]
df_respiratorio = df[ columnas ]

# Paso 2: Filtrar las filas por la condición de enfermedades respiratorias
# Códigos de causa de defunción que comienzan con 'J'
filtro_respiratorio = df_respiratorio['causa_def'].str.startswith('J')
df_respiratorio = df_respiratorio.loc[ filtro_respiratorio ]

# Muestra el nuevo DataFrame
df_respiratorio

,dia_ocurr,mes_ocurr,anio_ocur,ent_ocurr,mun_ocurr,causa_def,sexo,edad,edad_agru
0,25,11,2016,1,1,J441,2,4074,19
3,27,1,2017,1,1,J869,1,4080,21
17,2,11,2015,1,3,J841,1,4073,19
21,29,5,2016,1,1,J180,1,4049,14
32,24,1,2020,1,1,J449,1,4080,21
...,...,...,...,...,...,...,...,...,...
9597,3,8,2020,1,7,J189,1,3001,1
9937,6,9,2020,2,4,J189,1,4065,18
9976,30,8,2020,2,2,J449,1,4061,17
9989,23,12,2020,2,1,J988,1,4065,18


Después de seleccionar un subconjunto de datos, es crucial revisarlo para asegurar que la selección se hizo correctamente. Una primera revisión consiste en fijarse en el tamaño, el conjunto de datos original es de `10000 rows × 59 columns` mientras que el conjunto seleccionado es de `1041 rows × 9 columns`.

Esto nos sirve para verificar que las selecciones de filas y columnas se realizaron; sin embargo, es necesario revisar el **contenido** para validar la calidad de los datos. Estos hallazgos nos permiten diseñar una estrategia de limpieza para transformar los datos brutos en un formato confiable y listo para el análisis, la cual realizaremos en el siguiente notebook.

> Nota: recuerda que por limitaciones de recursos solo estamos trabajando las primeras 10000 filas.

## Ejercicio

Contesta la pregunta: ¿cuántos decesos se registraron en los municipios de Asientos, Calvillo, Pabellón Arteaga y Rincón de Romos agrupados por sexo y área rural o urbanas?

Utiliza las funciones vistas en este notebook para responder esto:
1. Determina las variables de interes
2. Selecciona el subconjunto de datos
3. Contesta la pregunta

Entrega un documento de texto con cada uno de los pasos, el código y las tablas resultantes. Puedes usar capturas de pantalla.

Los catálogos se encuentran en `conjunto_de_datos_defunciones_registradas_2020_csv/` catálogos.



. | area_urbana_rural |
----| --------|
1 | Urbana
2 | Rural
9 | No especificada
-

. | sexo |
----| --------|
1 | Hombre
2 | Mujer
9 | No especificado
-

. | municipio |
----| --------|
1* | Aguascalientes
2 | Asientos
3 | Calvillo
6 | Pabellón de Arteaga
7 | Rincón de Romos
* Clave de entidad

## Resumen

* Revisar un DataFrame con `.info()`
* Estructura de los conjuntos de datos ordenados o tidy data:
    * Cada observación forma una fila.
    * Cada variable forma una columna.
* Elementos de un dataframe: `.index`, `.columns`, `.values`, `.shape`, `.dtypes`.
* Seleccionar las primeras (`.head()`) y últimas (`.tail()`) filas.
* Seleccionar columnas: `df[ columnas ]`.
* Seleccionar por llaves: `df.loc[filas, columnas]`.
* Seleccionar por posición: `df.iloc[filas, columnas]`.
* Ordenar `.sort_values()` y eliminar `.drop()`.
* Evaluar si los valores de una columna cumplen una condición con: `==`, `!=`, `.isna()`, `.notna()`, `.isin()`.
* Mostrar las filas que cumplen una condición: `df[ df[col]==True ]`.
* Combinar múltiples condiciones con: `&`, `|` y `~`.
* Contar ocurrencias con: `.unique()`, `.nunique()` y `.value_counts()`.
* Restablecer el indice y volver el indice original en columnas con `.reset_index()`.

**¡Gracias!**

## Extra: Lectura de otros formatos de archivo

Las funciones de lectura en `pandas` son muy flexibles. La sintaxis cambia un poco dependiendo de la extensión del archivo.

  * **CSV (.csv)**: El formato más común para archivos de datos.
    ```python
    df_csv = pd.read_csv('mi_archivo.csv')
    ```
  * **Excel (.xlsx, .xls)**: Requiere la biblioteca `openpyxl` o `xlrd`.
    ```python
    df_excel = pd.read_excel('mi_archivo.xlsx')
    ```

## Extra: Errores comunes al abrir archivos

Durante la carga de datos, puedes encontrarte con algunos errores. Aquí te explicamos los más comunes y cómo solucionarlos.

**`UnicodeDecodeError` o caracteres extraños (`ñ`, `�`, etc)**. Este error ocurre cuando Python intenta leer un archivo de texto con la [codificación incorrecta](https://es.wikipedia.org/wiki/Codificaci%C3%B3n_de_caracteres), lo que suele suceder con los archivos que contienen caracteres especiales del español (como acentos o la letra 'ñ').

* Solución: La mayoría de las veces, basta con especificar el parámetro `encoding` en la función de lectura.
    ```python
    df = pd.read_csv('archivo_con_acentos.csv', encoding='latin1')
    ```

**`ImportError: Missing optional dependency`**. Este error significa que te falta una biblioteca opcional para leer un formato de archivo específico. Por ejemplo, `pandas` necesita `openpyxl` para leer archivos `.xlsx` de Excel.

* Solución: Instala la biblioteca que te indica el error usando `pip install`.
    ```bash
    !pip install openpyxl
    ```


## Extra: Principios de tidy data

Los principios de los conjuntos de datos ordenados (Tidy Data) fueron dados por Hadley Wickham. Recomendamos leer el [artículo original](http://vita.had.co.nz/papers/tidy-data.pdf), que traducimos parcialmente.

Los principios de los **datos ordenados** proporcionan una forma estándar de organizar los valores de datos dentro de un conjunto de datos. Un estándar hace que la limpieza inicial de datos sea más fácil, ya que no es necesario empezar desde cero y reinventar la rueda cada vez. El estándar de datos ordenados ha sido diseñado para facilitar la exploración inicial y el análisis de los datos, y para simplificar el desarrollo de herramientas de análisis de datos que funcionen bien juntas.

Al igual que las familias, los conjuntos de datos ordenados son todos similares, pero cada conjunto de datos desordenado es desordenado de su propia manera. Los conjuntos de datos ordenados proporcionan una forma estandarizada de vincular la estructura de un conjunto de datos (su disposición física) con su semántica (su significado). En esta sección, proporcionaré algunos términos estándar para describir la estructura y la semántica de un conjunto de datos, y luego utilizaré esas definiciones para definir los datos ordenados.

La mayoría de los conjuntos de datos estadísticos son tablas rectangulares compuestas por filas y columnas. Las columnas están casi siempre etiquetadas y las filas a veces están etiquetadas. Nuestro vocabulario de filas y columnas simplemente no es lo suficientemente rico como para describir por qué las dos tablas representan los mismos datos. Además de la apariencia, necesitamos una forma de describir la semántica subyacente o el significado de los valores mostrados en la tabla.

Un conjunto de datos es una colección de valores, generalmente números (si son cuantitativos) o cadenas (si son cualitativas). Los valores están organizados de dos maneras. Cada valor pertenece a una variable y una observación. Una variable contiene todos los valores que miden el mismo atributo subyacente (como la altura, la temperatura, la duración) a través de las unidades. Una observación contiene todos los valores medidos en la misma unidad (como una persona, un día o una carrera) a través de los atributos.

Los datos ordenados son una forma estándar de asignar el significado de un conjunto de datos a su estructura. Un conjunto de datos es
desordenado u ordenado dependiendo de cómo las filas, columnas y tablas se combinan con las observaciones,
variables y tipos. En datos ordenados:
* Cada observación forma una fila.
* Cada variable forma una columna.
* Cada tipo de unidad de observación forma una tabla.